packages/dht/src/dht/DhtNode.ts

```diff
@@ -0,0 +1,629 @@
+import { DhtPeer } from './DhtPeer'
+import KBucket from 'k-bucket'
+import PQueue from 'p-queue'
+import EventEmitter from 'events'
+import { SortedContactList } from './SortedContactList'
+import { RoutingRpcCommunicator } from '../transport/RoutingRpcCommunicator'
+import { ServerCallContext } from '@protobuf-ts/runtime-rpc'
+import { PeerID } from '../helpers/PeerID'
+import {
+    ClosestPeersRequest, ClosestPeersResponse,
+    ConnectivityResponseMessage,
+    Message,
+    NodeType,
+    PeerDescriptor, PingRequest, PingResponse, RouteMessageAck,
+    RouteMessageWrapper
+} from '../proto/DhtRpc'
+import { RouterDuplicateDetector } from './RouterDuplicateDetector'
+import * as Err from '../helpers/errors'
+import { ITransport, Event as ITransportEvent } from '../transport/ITransport'
+import { ConnectionManager } from '../connection/ConnectionManager'
+import { DhtRpcClient } from '../proto/DhtRpc.client'
+import { Logger } from '../helpers/Logger'
+import { v4 } from 'uuid'
+import { jsFormatPeerDescriptor } from '../helpers/common'
+import { IDhtRpc } from '../proto/DhtRpc.server'
+
+export interface RouteMessageParams {
+    message: Uint8Array
+    destinationPeer: PeerDescriptor
+    sourcePeer: PeerDescriptor
+    appId: string
+    previousPeer?: PeerDescriptor
+    messageId?: string
+}
+
+export enum Event {
+    NEW_CONTACT = 'streamr:dht:dht-node:new-peer',
+    CONTACT_REMOVED = 'streamr:dht:dht-node:peer-removed',
+    JOIN_COMPLETED = 'streamr:dht:dht-node:join-completed'
+}
+
+export const DEFAULT_APP_ID = 'layer0'
+
+export interface DhtNodeConfig {
+    transportLayer?: ITransport
+    peerDescriptor?: PeerDescriptor
+    entryPoints?: PeerDescriptor[]
+    webSocketHost?: string
+    webSocketPort?: number
+    peerIdString?: string
+    appId?: string
+    numberOfNodesPerKBucket?: number,
+    nodeName?: string
+}
+
+const logger = new Logger(module)
+
+export class DhtNode extends EventEmitter implements ITransport, IDhtRpc {
+    private noProgressCounter = 0
```

Comment:
What kind of progress this name refers to? (maybe could add a comment what this field tracks) Is it relevant only while `ongoingJoinOperation=true`?

Reply:
Add comment here

Reply:
Comment added